src/coreComponents/physicsSolvers/fluidFlow/wells/PerforationFluxKernels.hpp

```diff
@@ -0,0 +1,890 @@
+                          ElementSubRegionBase const & subRegion,
+                          MultiFluidBase const & fluid,
+                          CompFlowAccessors const & compFlowAccessors,
+                          MultiFluidAccessors const & multiFluidAccessors,
+                          RelPermAccessors const & relPermAccessors,
+                          bool const disableReservoirToWellFlow,
+                          ThermalCompFlowAccessors const & thermalCompFlowAccessors,
+                          ThermalMultiFluidAccessors const & thermalMultiFluidAccessors )
+    : Base( perforationData,
+            subRegion,
+            compFlowAccessors,
+            multiFluidAccessors,
+            relPermAccessors,
+            disableReservoirToWellFlow ),
+    m_stack( perforationData,
+             fluid,
+             thermalCompFlowAccessors,
+             thermalMultiFluidAccessors )
+
+  {}
+
+
+  struct StackVariables : public Base::StackVariables
+  {
+public:
+
+    GEOS_HOST_DEVICE
+    StackVariables( PerforationData * const perforationData,
+                    MultiFluidBase const & fluid,
+                    ThermalCompFlowAccessors const & thermalCompFlowAccessors,
+                    ThermalMultiFluidAccessors const & thermalMultiFluidAccessors )
+      : Base::StackVariables(),
+                                       m_wellElemPhaseFrac( fluid.phaseFraction() ),
+      m_dPhaseFrac( fluid.dPhaseFraction() ),
+      m_wellElemPhaseEnthalpy( fluid.phaseEnthalpy()),
+      m_dWellElemPhaseEnthalpy( fluid.dPhaseEnthalpy()),
+      m_energyPerfFlux( perforationData->getField< fields::well::energyPerforationFlux >()),
+      m_dEnergyPerfFlux( perforationData->getField< fields::well::dEnergyPerforationFlux >()),
+      m_temp( thermalCompFlowAccessors.get( fields::flow::temperature {} ) ),
+      m_resPhaseEnthalpy( thermalMultiFluidAccessors.get( fields::multifluid::phaseEnthalpy {} ) ),
+      m_dResPhaseEnthalpy( thermalMultiFluidAccessors.get( fields::multifluid::dPhaseEnthalpy {} ) )
+    {}
+
+    /// Views on well element properties
+    /// Element phase fraction
+    arrayView3d< real64 const, multifluid::USD_PHASE > const m_wellElemPhaseFrac;
+    arrayView4d< real64 const, multifluid::USD_PHASE_DC > const m_dPhaseFrac;
+    arrayView3d< real64 const, multifluid::USD_PHASE >  m_wellElemPhaseEnthalpy;
+    arrayView4d< real64 const, multifluid::USD_PHASE_DC >  m_dWellElemPhaseEnthalpy;
+
+    /// Views on energy flux
+    arrayView1d< real64 > const & m_energyPerfFlux;
+    arrayView3d< real64 > const & m_dEnergyPerfFlux;
+
+    /// Views on temperature
+    ElementViewConst< arrayView1d< real64 const > > const m_temp;
+
+    /// Views on phase enthalpies
+    ElementViewConst< arrayView3d< real64 const, multifluid::USD_PHASE > > const m_resPhaseEnthalpy;
+    ElementViewConst< arrayView4d< real64 const, multifluid::USD_PHASE_DC > > const m_dResPhaseEnthalpy;
```

Comment:
StackVariables was used to support constexpr for thermal parts in isothermal compute.  Not sure if it can be removed, unless dispatch is used and constexpr is removed.   I like the inlining that constexpr gives you... I'll try initializing the stack every time in the icon loop as in res flux,  originally thought this is overhead so did it on class construction

Reply:
oh, I see. It can't really be done like that. So, yes you will have to either use that callback design to assemble the two cases. 
If we think that it's nice to have a single `computePerforationFlux` function with an `if constexpr `block to separate the two cases, we could do something like this https://godbolt.org/z/Ehhh5o983 but I would like to hear what @wrtobin and @rrsettgast think about this before sending you down this path.

Reply:
The constexpr was introduced and discussed early in dev.   A fair bit of code will need to be added/refactored if this isnt going to be allowed.